### PR TITLE
Require Kokkos 3.7

### PR DESCRIPTION
**Pull request by aprokop**
Updates Kokkos versions in several builds:

| Build | Old | New |
| - | - | - |
| GitLab | 3.6.00 | 3.7.00 |
| `HIP-5.0` | 3.6.00 | 4.0.00 |
| `GCC-12.2` | 3.6.00 | 4.0.00 |
| `Clang` | 3.6.00 | 4.0.00 |
| `CUDA-11.5.2-NVCC-CUDA-AWARE-MPI`| 3.6.00 | 3.7.00 |
| `CUDA-11.5.2-NVCC` | 3.6.00 | 4.0.00 |

### CODE REVIEW

File: benchmarks/brute_force_vs_bvh/brute_force_vs_bvh.cpp [hunk 1]
```diff
@@ -53,7 +53,6 @@ int main(int argc, char *argv[])
   case 3:
     run<3>(nprimitives, nqueries, nrepeats);
     break;
-#if KOKKOS_VERSION >= 30700
   case 2:
     run<2>(nprimitives, nqueries, nrepeats);
     break;
```
Comment (dalg24) [inline benchmarks/brute_force_vs_bvh/brute_force_vs_bvh.cpp:58]: Probably would not hurt to reorder 2<->3

File: benchmarks/dbscan/dbscan.cpp [hunk 2]
```diff
@@ -149,7 +149,6 @@ int main(int argc, char *argv[])
   case 3:
     success = run<3>(params);
     break;
-#if KOKKOS_VERSION >= 30700
   case 2:
     success = run<2>(params);
     break;
```
Comment (dalg24) [inline benchmarks/dbscan/dbscan.cpp:154]: Same comment about ordering

File: .jenkins/continuous.groovy [hunk 3]
```diff
@@ -42,7 +42,7 @@ pipeline {
                         dockerfile {
                             filename "Dockerfile"
                             dir "docker"
-                            additionalBuildArgs '--build-arg BASE=nvidia/cuda:11.5.2-devel-ubuntu20.04 --build-arg KOKKOS_OPTIONS="-DCMAKE_CXX_EXTENSIONS=OFF -DKokkos_ENABLE_SERIAL=ON -DKokkos_ENABLE_CUDA=ON -DKokkos_ENABLE_CUDA_LAMBDA=ON -DKokkos_ARCH_VOLTA70=ON" --build-arg CUDA_AWARE_MPI=1'
+                            additionalBuildArgs '--build-arg BASE=nvidia/cuda:11.5.2-devel-ubuntu20.04 --build-arg KOKKOS_VERSION=3.7.00 --build-arg KOKKOS_OPTIONS="-DCMAKE_CXX_EXTENSIONS=OFF -DKokkos_ENABLE_SERIAL=ON -DKokkos_ENABLE_CUDA=ON -DKokkos_ENABLE_CUDA_LAMBDA=ON -DKokkos_ARCH_VOLTA70=ON" --build-arg CUDA_AWARE_MPI=1'
```
Comment (dalg24) [inline .jenkins/continuous.groovy:45]: please clearly state in the description what are the resulting changes in the kokkos version of CI builds

Comment (aprokop) [reply .jenkins/continuous.groovy:45]: Done.

**Review by masterleinad — APPROVED** [review]
Looks OK to me.

**Comment by dalg24** [comment]
Did you mean to move the majority of builds to 4.0?

**Comment by aprokop** [comment]
> Did you mean to move the majority of builds to 4.0?

Yes. There are now 4 Kokkos 4.0 builds, and 3 Kokkos 3.7 builds. I think that's a reasonable split.

**Review by dalg24 — CHANGES_REQUESTED** [review]
Kokkos 4 requires a minimum ROCm version of 5.2
So you'd need to update the base image

**Comment by aprokop** [comment]
HIP fails with
```
CMake Error at cmake/kokkos_compiler_id.cmake:194 (MESSAGE):
  Compiler not supported by Kokkos.  Required compiler versions:

      Clang(CPU)         8.0.0 or higher
      Clang(CUDA)       10.0.0 or higher
      GCC                8.2.0 or higher
      Intel             19.0.5 or higher
      IntelLLVM(CPU)  2021.1.1 or higher
      IntelLLVM(SYCL) 2022.0.0 or higher
      NVCC              11.0.0 or higher
      HIPCC              5.2.0 or higher
      NVHPC/PGI           22.3 or higher
      MSVC               19.29 or higher
      XL/XLClang           not supported

  Compiler: HIPCC 5.0.13601
```
Updated HIP container to use 5.2. Could go higher, to 5.3.3 or 5.4.2 if we want.

**Comment by aprokop** [comment]
HIP error:
```
fatal error: error in backend: Cannot select: 0x559739e39148: i32,ch = 
AtomicLoadAdd<(load store syncscope("agent-one-as") monotonic (s32) on %ir.161, 
addrspace 5)> 0x5597327b0b68, 0x55972efc81a8, Constant:i32<1>, /opt/rocm-
5.2.0/include/hip/amd_detail/amd_hip_atomic.h:141:10 @[ /opt/kokkos/include/desul
/atomics/Fetch_Op_HIP.hpp:48:159 @[ /opt/kokkos/include/desul/atomics
/Generic.hpp:327:3 @[ /opt/kokkos/include/Kokkos_Atomics_Desul_Wrapper.hpp:121:44 
@[ test/tstQueryTreeRay.cpp:146:28 @[ src/details
/ArborX_DetailsTreeTraversal.hpp:438:7 @[ /opt/kokkos/include
/HIP/Kokkos_HIP_Parallel_Range.hpp:58:5 @[ /opt/kokkos/include
/HIP/Kokkos_HIP_Parallel_Range.hpp:73:22 @[ /opt/kokkos/include
/HIP/Kokkos_HIP_KernelLaunch.hpp:80:3 ] ] ] ] ] ] ] ]
  0x55972efc81a8: i32 = add 0x55973360e200, Constant:i32<104>
    0x55973360e200: i32 = add FrameIndex:i32<0>, 0x55972c2dc648
      0x5597348aba68: i32 = FrameIndex<0>
      0x55972c2dc648: i32 = shl 0x559735c47df0, Constant:i32<2>
        0x559735c47df0: i32 = truncate 0x5597285b60a0
          0x5597285b60a0: i64 = AssertSext 0x559735c47d88, ValueType:ch:i32
            0x559735c47d88: i64,ch = CopyFromReg 0x5597327b0b68, Register:i64 %11
              0x55972efc8b00: i64 = Register %11
        0x55972c2dc8b8: i32 = Constant<2>
    0x55973360de58: i32 = Constant<104>
  0x55972b981710: i32 = Constant<1>
In function: _ZN6Kokkos4ImplL32hip_parallel_launch_local_memoryINS0_11ParallelForIN6ArborX7Details13TreeTraversalINS3_28BasicBoundingVolumeHierarchyINS_8HIPSpaceENS3_3BoxEvEENS4_12PermutedDataI28BoxesIntersectedByRayOrderedINS_6DeviceINS_3HIPES7_EEENS_4ViewIPjJSE_EEELb0EEE19InsertIntersectionsISE_ENS3_12Experimental26OrderedSpatialPredicateTagEEENS_11RangePolicyIJSD_NSO_11OneLeafTreeEEEESD_EELj1024ELj1EEEvT_
```

**Comment by aprokop** [comment]
HIP 5.4.2 passes.

**Comment by aprokop** [comment]
HIP 5.3.3 also works. Merging.

**Comment by aprokop** [comment]
Merged with different builds passing at certain points (but not all together).

**Pull request merged**